### PR TITLE
A workable version for Softlayer

### PR DESCRIPTION
**Pull request by mattcui**
Already verified with `zookeeper` release.

/cc @maximilien @jianqiu

### CODE REVIEW

File: softlayer/cpi.yml [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+  value: *cpi_job
+
+- type: replace
+  path: /instance_groups/name=bosh/properties/softlayer?
+  value: &softlayer
+    username: ((sl_username))
+    apiKey: ((sl_api_key))
+    featureOptions:
+      apiEndpoint: api.service.softlayer.com
+      apiRetryCount: 10
```
Comment (cppforlife) [inline softlayer/cpi.yml]: why arent these property values default on the cpi?

File: softlayer/cpi.yml [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+- type: replace
+  path: /cloud_provider/template?
+  value: *cpi_job
+
+- type: replace
+  path: /instance_groups/name=bosh/properties/softlayer?
+  value: &softlayer
+    username: ((sl_username))
+    apiKey: ((sl_api_key))
```
Comment (cppforlife) [inline softlayer/cpi.yml]: keys should be snake case.

File: softlayer/cloud-config.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+- name: default
+  cloud_properties:
+    bosh_ip: ((internal_ip))
+    startCpus:  4
+    maxMemory:  8192
+    ephemeralDiskSize: 100
+    hourlyBillingFlag: true
+    vmNamePrefix: ((sl_vm_name_prefix))
+    domain: ((sl_vm_domain))
+    datacenter:
```
Comment (cppforlife) [inline softlayer/cloud-config.yml]: datacenter is already specified on the az, so no need to specify on vm type.

File: softlayer/cloud-config.yml [hunk 4]
```diff
@@ -0,0 +1,51 @@
+azs:
+- name: z1
```
Comment (cppforlife) [inline softlayer/cloud-config.yml]: should have 3 azs even if as a starter they are all the same. makes default example manifests work much better.

File: softlayer/cloud-config.yml [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+  cloud_properties:
+    datacenter: { name: ((sl_datacenter)) }
+
+vm_types:
+- name: default
+  cloud_properties:
+    bosh_ip: ((internal_ip))
```
Comment (cppforlife) [inline softlayer/cloud-config.yml]: why do we need internal_ip over here?

File: softlayer/cloud-config.yml [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    startCpus:  4
+    maxMemory:  8192
+    ephemeralDiskSize: 100
+    hourlyBillingFlag: true
+    vmNamePrefix: ((sl_vm_name_prefix))
+    domain: ((sl_vm_domain))
+    datacenter:
+      name: ((sl_datacenter))
+
+disk_types:
```
Comment (cppforlife) [inline softlayer/cloud-config.yml]: should have a disk type called `default`

File: softlayer/cpi.yml [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+# Configure sizes
+- type: replace
+  path: /resource_pools/name=vms/cloud_properties?
+  value:
+    domain: ((sl_vm_domain))
+    vmNamePrefix: ((sl_vm_name_prefix))
+    ephemeralDiskSize: 100
+    startCpus: 4
+    maxMemory: 8192
+    deployedByBoshcli: true
```
Comment (cppforlife) [inline softlayer/cpi.yml]: why do we need `deployedByBoshcli`

**Comment by cppforlife** [comment]
@mattcui left some comments.

File: softlayer/cpi.yml (hunk 1 above)
Comment (mattcui) [inline softlayer/cpi.yml]: By default, it uses public SL API endpoint (api.softlayer.com), I just hope to show there is an option to change to use the private SL API endpoint, I could delete them if you don't think it's necessary.

File: softlayer/cpi.yml (hunk 2 above)
Comment (mattcui) [inline softlayer/cpi.yml]: OK, I will modify it. Thanks.

File: softlayer/cloud-config.yml (hunk 3 above)
Comment (mattcui) [inline softlayer/cloud-config.yml]: OK, I will have test to make sure our CPI works after removing `datacenter` from `vm_types`

File: softlayer/cloud-config.yml (hunk 4 above)
Comment (mattcui) [inline softlayer/cloud-config.yml]: OK, I will add more for Dal and SJ. Thanks.

File: softlayer/cloud-config.yml (hunk 5 above)
Comment (mattcui) [inline softlayer/cloud-config.yml]: It's used to combine mbus url in user_data.json, we are already able to set the private IP for dynamic networking, but haven't implemented it for static networking.

File: softlayer/cloud-config.yml (hunk 6 above)
Comment (mattcui) [inline softlayer/cloud-config.yml]: Yes, I will modify. Thanks.

File: softlayer/cpi.yml (hunk 7 above)
Comment (mattcui) [inline softlayer/cpi.yml]: Actually, it's used to get ride of `bosh_ip` from each resource pools (in manifest v1), for deploying bosh director, there needs to update /etc/hosts for supporting dynamic IP. Below is the code where to use it.

https://github.com/cloudfoundry/bosh-softlayer-cpi-release/blob/master/src/bosh-softlayer-cpi/softlayer/vm/softlayer_virtual_guest_creator.go#L103-L130

File: softlayer/cpi.yml (hunk 1 above)
Comment (cppforlife) [inline softlayer/cpi.yml]: > I just hope to show there is an option to change to use the private SL API endpoint, I could delete them if you don't think it's necessary.

that should be covered by docs & https://bosh.io/jobs/softlayer_cpi?source=github.com/cloudfoundry/bosh-softlayer-cpi-release

Comment (mattcui) [inline softlayer/cpi.yml]: OK, I am deleting them, and yes, it's already covered in https://bosh.io/jobs/softlayer_cpi?source=github.com/cloudfoundry/bosh-softlayer-cpi-release

**Comment by mattcui** [comment]
@cppforlife I addressed your comments, please review again, thank you. For comments you said there left, do you mean the comments like `# Add CPI job`, I saw it's in cpi.yml of all other IaaSes, do I need to removed all comments in cpi.yml? Thanks.

File: softlayer/cpi.yml [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+
+- type: replace
+  path: /cloud_provider/template?
+  value: *cpi_job
+
+- type: replace
+  path: /instance_groups/name=bosh/properties/softlayer?
+  value: &softlayer
+    username: ((sl_username))
+    apiKey: ((sl_apiKey))
```
Comment (cppforlife) [inline softlayer/cpi.yml]: this one seems to still have mixed casing case.

**Comment by cppforlife** [comment]
i meant comments on github =)

**Comment by mattcui** [comment]
@cppforlife I changed `sl_apiKey` to `sl_api_key`, please let me know if any other things I need to change. Thanks.

**Comment by cppforlife** [comment]
I merged this PR; however, let's add softlayer config to `test.sh`. Also am i understanding correctly that /etc/hosts update isnt nececssary anymore?

**Pull request closed without merging**